File: app/services/provider_registry.py
from typing import Generic, TypeVar

T = TypeVar("T")


class ProviderRegistry(Generic[T]):
    def __init__(self, *, default_provider: str | None = None):
        self._default_provider = self._normalize(default_provider) if default_provider else None
        self._providers: dict[str, T] = {}

    @staticmethod
    def _normalize(name: str | None) -> str:
        return str(name or "").strip().lower()

    def register(self, name: str, provider: T) -> None:
        key = self._normalize(name)
        if not key:
            return
        self._providers[key] = provider

    def get(self, name: str) -> T | None:
        key = self._normalize(name)
        return self._providers.get(key)

    def resolve(self, name: str) -> T:
        provider = self.get(name)
        if provider is not None:
            return provider

        if self._default_provider:
            default = self._providers.get(self._default_provider)
            if default is not None:
                return default

        raise KeyError(name)

File: backend-ts/src/services/provider-registry.ts
/**
 * Generic provider registry
 * Migrated from Python: app/services/provider_registry.py
 *
 * Strategy/registry pattern for pluggable providers (LLM generators,
 * publishers, safety checkers, search providers, etc.).
 */

export class ProviderRegistry<T> {
  private readonly providers = new Map<string, T>();
  private readonly defaultProvider: string | null;

  constructor(options?: { defaultProvider?: string }) {
    const raw = options?.defaultProvider;
    this.defaultProvider = raw ? normalize(raw) : null;
  }

  register(name: string, provider: T): void {
    const key = normalize(name);
    if (!key) return;
    this.providers.set(key, provider);
  }

  get(name: string): T | undefined {
    return this.providers.get(normalize(name));
  }

  resolve(name: string): T {
    const provider = this.get(name);
    if (provider !== undefined) return provider;

    if (this.defaultProvider) {
      const fallback = this.providers.get(this.defaultProvider);
      if (fallback !== undefined) return fallback;
    }

    throw new Error(`Provider not found: "${name}"`);
  }

  get size(): number {
    return this.providers.size;
  }

  keys(): string[] {
    return [...this.providers.keys()];
  }
}

function normalize(name: string | undefined | null): string {
  return (name ?? '').trim().toLowerCase();
}
